lib/pbench/test/unit/server/query_apis/test_index_mappings.py

```diff
@@ -0,0 +1,62 @@
+import pytest
+from http import HTTPStatus
+from sqlalchemy.exc import DatabaseError
+
+from pbench.server.database.models.template import Template
+
+
+@pytest.fixture()
+def database_error(monkeypatch):
+    def raise_db_error(name: str):
+        raise DatabaseError("DB Error")
+
+    with monkeypatch.context() as m:
+        m.setattr(Template, "find", raise_db_error)
+        yield
+
+
+class TestIndexMappings:
+    """
+    Unit testing for resources/IndexMappings class.
+
+    In a web service context, we access class functions mostly via the
+    Flask test client rather than trying to directly invoke the class
+    constructor and `post` service.
+    """
+
+    def test_query(self, client, server_config, find_template):
+        """
+        Check the construction of index mappings API and filtering of the
+        response body.
+        """
+        with client:
+            response = client.get(f"{server_config.rest_uri}/index/mappings")
+            assert response.status_code == HTTPStatus.OK
+            res_json = response.json
+            assert isinstance(res_json, dict)
+            assert len(res_json.keys()) == 5
+            assert list(res_json.keys()) == [
+                "@metadata",
+                "authorization",
+                "host_tools_info",
+                "run",
+                "sosreports",
+            ]
+
+    def test_with_no_run_document(self, client, server_config):
+        """
+        Check the index mappings API if there is no run document present in the database.
+        """
+        with client:
+            response = client.get(f"{server_config.rest_uri}/index/mappings")
+            assert response.status_code == HTTPStatus.INTERNAL_SERVER_ERROR
+            assert response.json["message"] == "INTERNAL ERROR"
```

Comment:
This is a curious case.  Is it really true that there should never be a server deployment with no run document present in the database?  (So, part of server installation is to initialize the database with a run document?)

If so, this feels more like something which should be an assertion in the Server code or in an installation test, not a unit test.

Reply:
This is an interesting question, I also believe this to be a far extreme edge case. As far as I know when someone moves their data on the server (and considering there is no previous data on the server) the prep-shim picks up the data and the template database gets created on the server which has the run document index in it. 
@dbutenhof any thought?

Reply:
Right now, it *is* possible, although it's a startup edge case we'll probably want to close at some point.

The template DB table is filled in by the *first* call to the indexer, which means the first time a dataset is received via `PUT` or staged by `prep-shim`, gets unpacked, and marked for indexing. (Actually, it may be a bit better than that; server tools generate `server-report` documents even when they don't do anything, and I think the report generator code probably loads *all* of the templates; and the first tool run should happen pretty quickly after `crontab` gets started.)

Of course it's possible to connect a dashboard (or other client) and make queries on the active API before that. They won't return anything interesting, at best, but we should be robust.

One option would be to have something during install/configure run through the `PbenchTemplates` procedure before we open the server for business, so that it's populated. Another option would be for the API code to just fail with some meaningful message (sorry, waiting for inventory) until we actually index the first dataset.